zenoh-ext/src/net/querying_subscriber.rs

```diff
@@ -0,0 +1,483 @@
+        self
+    }
+}
+
+impl<'a> Future for QueryingSubscriberBuilder<'a> {
+    type Output = ZResult<QueryingSubscriber<'a>>;
+
+    fn poll(self: Pin<&mut Self>, _cx: &mut Context<'_>) -> Poll<Self::Output> {
+        Poll::Ready(QueryingSubscriber::new(Pin::into_inner(self).clone()))
+    }
+}
+
+impl<'a> ZFuture<ZResult<QueryingSubscriber<'a>>> for QueryingSubscriberBuilder<'a> {
+    fn wait(self) -> ZResult<QueryingSubscriber<'a>> {
+        QueryingSubscriber::new(self)
+    }
+}
+
+pub struct QueryingSubscriber<'a> {
+    conf: QueryingSubscriberBuilder<'a>,
+    subscriber: Subscriber<'a>,
+    receiver: QueryingSubscriberReceiver,
+}
+
+impl QueryingSubscriber<'_> {
+    fn new(conf: QueryingSubscriberBuilder<'_>) -> ZResult<QueryingSubscriber<'_>> {
+        // declare subscriber at first
+        let mut subscriber = conf
+            .session
+            .declare_subscriber(&conf.sub_reskey, &conf.info)
+            .wait()?;
+
+        let receiver = QueryingSubscriberReceiver::new(subscriber.receiver().clone());
+
+        let mut query_subscriber = QueryingSubscriber {
+            conf,
+            subscriber,
+            receiver,
+        };
+
+        // start query
+        query_subscriber.query()?;
+
+        Ok(query_subscriber)
+    }
+
+    /// Undeclare this QueryingSubscriber
+    #[inline]
+    pub fn undeclare(self) -> ZResolvedFuture<ZResult<()>> {
+        self.subscriber.undeclare()
+    }
+
+    /// Return the QueryingSubscriberReceiver associated to this subscriber.
+    #[inline]
+    pub fn receiver(&mut self) -> &mut QueryingSubscriberReceiver {
+        &mut self.receiver
+    }
+
+    /// Issue a new query using the configured resource key and predicate.
+    pub fn query(&mut self) -> ZResult<()> {
+        self.query_on(
+            &self.conf.query_reskey.clone(),
+            &self.conf.query_predicate.clone(),
+            self.conf.query_target.clone(),
+            self.conf.query_consolidation.clone(),
+        )
+    }
+
+    /// Issue a new query on the specified resource key and predicate.
+    pub fn query_on(
+        &mut self,
+        reskey: &ResKey,
+        predicate: &str,
+        target: QueryTarget,
+        consolidation: QueryConsolidation,
+    ) -> ZResult<()> {
+        let mut state = zwrite!(self.receiver.state);
+
+        if state.query_replies_recv.is_none() {
+            log::debug!("Start query on {}?{}", reskey, predicate);
+            state.query_replies_recv = Some(
+                self.conf
+                    .session
+                    .query(reskey, predicate, target, consolidation)
+                    .wait()?,
+            );
+            Ok(())
+        } else {
+            log::error!(
+                "Cannot start query on {}?{} - one is already in progress",
+                reskey,
+                predicate
+            );
+            zerror!(ZErrorKind::Other {
+                descr: "Query already in progress".to_string()
+            })
```

Comment:
Do we really want to return an error here ? The user has no clue if a query is already pending or not when calling query(). Returning an error when the user did "nothing bad" and have no control on it seems wrong to me.

Reply:
Sure. I'll change the implementation to support several queries at the same time. Thus no error will be returned, but the new query will be added to an inner list of queries in progress.